### PR TITLE
[BEAM-6597] Replace ProgressRequestCallback with BundleProgressReporter

### PR DESCRIPTION
**Pull request by lukecwik**
This cleans up a TODO for a closed issue and moves short id logic to one time instantiation vs per bundle report for a minor perf improvement.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lukecwik** [comment]
This is dependent on https://github.com/apache/beam/pull/22002 being merged as a precursor to this.

CC: @apilloud @TheNeuralBit

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Java PreCommit

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java [hunk 1]
```diff
@@ -1021,13 +991,12 @@ void shutdown() throws Exception {
   public abstract static class BundleProcessor {
     public static BundleProcessor create(
         Cache<Object, Object> processWideCache,
-        InMemory bundleProgressReporterAndRegistrar,
+        BundleProgressReporter.InMemory bundleProgressReporterAndRegistrar,
```
Comment (Abacn) [inline sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java:994]: I see, there are two "InMemory" which are subclass of different interface. If so then may remove the import of `org.apache.beam.fn.harness.control.BundleProgressReporter.InMemory` and also change all other raw InMemory to `BundleProgressReporter.InMemory` in this file

**Review by Abacn — COMMENTED** [review]
I am not familiar with harness. Went through once and just find a minor piece. Just curious, is the expected performance gain evaluated in some way (like [here](https://github.com/apache/beam/pull/22002)). Thanks!

**Comment by lukecwik** [comment]
> I am not familiar with harness. Went through once and just find a minor piece. Just curious, is the expected performance gain evaluated in some way (like [here](https://github.com/apache/beam/pull/22002)). Thanks!

It should help some but not a meaningful amount as this isn't happening on the element processing hot path.

File: sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java (hunk 1 above)
Comment (lukecwik) [reply sdks/java/harness/src/main/java/org/apache/beam/fn/harness/control/ProcessBundleHandler.java:994]: Thanks I missed this one.

Fixed.

**Comment by lukecwik** [comment]
@Abacn PTAL

**Review by Abacn — APPROVED** [review]
LGTM to the best of my knowledge. May find someone have a look on the fn runner change.

**Pull request merged**